### PR TITLE
Added support for emoticons

### PR DESCRIPTION
**Pull request by moe-25**
This adds support for emoticons resp. characters above 0xFFFF in general. It might be partially related to issue #145. Since PDF cannot handle Unicode surrogates directly, they have to be re-mapped to fit 16-bit. For simplicity, each character is mapped the first time it is used, allowing PDF documents with 655535 __different__ characters per font. Even if glyphs are not present in the font, they can be made visible by copy-and-paste. These changes could probably break other things (including various tests).

* `text()` and `cell()` must remap all characters in incoming strings
* `TTFontFile.makeSubset()` must be informed where to place each glyph in the embedded font
* `TTFontFile` must consider CMAP12 tables to actually reach these glyphs
* Remapping must be taken into account when creating the charWidth definition
* Instead of _bfrange_, _bfchar_ is used to support PDF readers when searching/copying.

There would still be a need for optimization, such as replacing the `subset` array with a dictionary, shrinking the subset when `alias_nb_pages` is used and so on. How do you think about this approach?

Demo:
```
x = fpdf.FPDF()
x.add_font('DejaVuSans', '', 'DejaVuSans.ttf', uni=True)
x.set_font('DejaVuSans')
x.add_page()

x.cell(0, None, "Page {nb}", align="C", ln=1)

x.cell(50, 50, "That's copyable \U0001F5B1")
x.text(100, 50, "That's cool \U0001F60E")

x.output('Test.pdf')
```

### CODE REVIEW

File: fpdf/fpdf.py [hunk 1]
```diff
@@ -1066,7 +1066,14 @@ def add_font(self, family, style="", fname=None, uni=False):
                 unifilename = cache_dir / f"{ttffilename.stem}.pkl"
 
             # include numbers in the subset! (if alias present)
-            sbarr = list(range(57 if self.str_alias_nb_pages else 32))
+            # ensure that alias is mapped 1-by-1 additionally (must be replaceable)
+            if self.str_alias_nb_pages:
+                subset_max = 57
+                for char in self.str_alias_nb_pages:
+                    subset_max = max(subset_max, ord(char))
```
Comment (Lucas-C) [inline fpdf/fpdf.py:1073]: Why this change?
In practice, `sbarr` wille become `list(range(125))` if `str_alias_nb_pages` is set.

I still wonder where this value of `57` is coming from...

Also, it terms of code, this would be shorter:
```python
subset_max = max(map(ord, self.str_alias_nb_pages))
```

File: fpdf/ttfonts.py [hunk 2]
```diff
@@ -414,7 +418,7 @@ def extractInfo(self):
         # hmtx - Horizontal metrics table
         self.getHMTX(numberOfHMetrics, numGlyphs, glyphToChar, scale)
 
-    def makeSubset(self, file, subset):
+    def makeSubset(self, file, subset, remap = {}):
```
Comment (Lucas-C) [inline fpdf/ttfonts.py:421]: Could you maybe add a short docstring explaining what is the role of this `remap` dict,
and what are its keys & values?

Also, please do not set a mutable default value as it can be very error prone:
https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

**Review by Lucas-C — CHANGES_REQUESTED** [review]
Wow, this looks really promising!
Thank you for tackling this subject 🙏

Could you please add some tests that produce PDF files,
to get an idea of what users will be able to produce?
_cf._ https://pyfpdf.github.io/fpdf2/Development.html#testing

**Comment by Lucas-C** [comment]
@all-contributors please add @moe-25 for research!

**Comment by Lucas-C** [comment]
@all-contributors please add @moe-25 for research !

**Comment by Lucas-C** [comment]
@all-contributors please add @moe-25 for research

File: fpdf/fpdf.py (hunk 1 above)
Comment (moe-25) [reply fpdf/fpdf.py:1073]: Substituting page numbers became more complicated. When `_substitute_page_number()` calls `str.replace()`, it is no longer known which font/mapping was used. For example, if a dedicated font is used for the page numbers, _{nb}_ would become _\x01\x02\x03\x04_. The same string is created if another text starts with four different characters (which is likely), which would be replaced as well then. The simplest solution to prevent conflicts is that all characters of the `str_alias_nb_pages` and its resulting digits always get the same mapping.

To avoid having to embed all 125 characters, `font["subset"]` should be of another type. I guess `dict.get()` is also faster than `list.index()` 😉 I'll give it a try.

Thanks for your feedback!

Cheers,
Moe

**Review by alexp1917 — APPROVED** [review]
given that this can be formatted and made to pass tests, looks great. while i would like for there to be many more comments on clarify as to what is going on, i dont get it now, and if i dont after this PR, this changes nothing for me.

part of the mystery is the problem is the (nonexistant) code structure. so if the explanation of the code actually comes with a suggestion on how to re-organize it too, thats fine by me. thanks for the contribution! neat stuff

**Comment by alexp1917** [comment]
and yes, tests are important, but tests exist to test some detail in the implementation, which as far as I'm concerned, is a pretty untestable black box.

**Comment by moe-25** [comment]
I updated the branch to include the discussed aspects.

Most notably, `font['subset']` is no longer a list but instances a new class (that's why I noticed it shouldn't be pickled in #187). I'm pretty sure that `SubsetMap` can be optimized but it shouldn't make things worse (earlier, it stored a copy of the entire text because every character was appended regardless of its previous existence). I also added some demos/tests to print some DejaVuSans glyphs with unicodes higher 0xFFFF.

Finally, we might discuss whether to change `TTFontFile.subset`'s signature. My first though was just adding a new parameters, but now I think changing `subset` to actually be the mapping dictionary is more suitable. The current version auto-detects the given type.

Happy to hear your opinion again. Awesome library!

File: fpdf/fpdf.py [hunk 3]
```diff
@@ -121,6 +121,43 @@ class ToCPlaceholder(NamedTuple):
...
+    tries to return the lowest possible representation.
+    """
+
+    def __init__(self, identities: List[int]):
+        super().__init__()
+        self._next = 0
+
+        # sort list to ease deletion once _next
+        # becomes higher than first reservation
+        self._reserved = sorted(identities)
```
Comment (Lucas-C) [inline fpdf/fpdf.py:140]: This really is an edge case, but `SubsetMap.pick` will no behave as expected if `identities` contain duplicates here, _e.g._ if `identities == (0, 0, 0, 2)`

File: fpdf/fpdf.py [hunk 4]
```diff
@@ -2652,6 +2710,24 @@ def _putfonts(self):
...
+                # bfChar
+                # This table informs the PDF reader about the unicode
+                # character that each used 16-bit code belongs to. It
+                # allows searching the file and copying text from it.
+                bfChar = []
+                subset = font["subset"].dict()
+                for code in subset:
+                    code_mapped = subset.get(code)
+                    if code > 0xFFFF:
+                        # Calculate suggorate pair
```
Comment (Lucas-C) [inline fpdf/fpdf.py:2722]: Minor typo: "surrogate"

**Review by Lucas-C — APPROVED** [review]

File: test/fonts/test_font_remap.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
+from pathlib import Path
```
Comment (Lucas-C) [inline test/fonts/test_font_remap.py:1]: Really great tests, good job! 👍

File: fpdf/fpdf.py (hunk 3 above)
Comment (moe-25) [reply fpdf/fpdf.py:140]: Indeed, I have not considered this case. The same applies to negative values. But apart from the fact that `_reserved` would not be cleared properly, it should actually have no noticeable impact. I will now point this out in the docstring and weakly type `identities` as `Set[int]`. `sbarr` will only hold acceptable values in practice anyway. Do you think this is enough or should it still be checked more aggressively?

**Comment by moe-25** [comment]
The mentioned change in `SubsetMap` and a bugfix in the surrogate calculation are still pending.

I also found a bug in `text()` that obviously didn't come into play before. Since a separate PR is not worthwhile, I would fix that right away.

Demo against master:
```
pdf = FPDF()
pdf.set_compression(False)
pdf.add_font('DejaVuSans', '', 'DejaVuSans.ttf', uni=True)
pdf.set_font('DejaVuSans')
pdf.add_page()
pdf.text(50, 50, 'Te()st') # Parentheses won't be escaped correctly
pdf.output('Demo.pdf')
```

This feature now seems to be complete from my point of view. If you agree, I'd be happy to rebase everything.

Cheers,
Moe

File: fpdf/fpdf.py (hunk 3 above)
Comment (Lucas-C) [reply fpdf/fpdf.py:140]: That's perfect for me, no need for more defensive programming here I think 😉

**Comment by Lucas-C** [comment]
Merged!

Thanks a lot @moe-25 😉

**Comment by moe-25** [comment]
> The mentioned change in `SubsetMap` and a bugfix in the surrogate calculation are still pending.

No wait... :fearful: There were still bugs I wanted to fix today. To be honest, I don't know if I can just push them into my branch now or if it requires a new PR. Can you help me?

**Comment by Lucas-C** [comment]
Wooops, sorry, my mistake!

I thought this was ready to be merged 🤭

No worries though, I haven't made a new release yet.

It would be best to open a new PR now, but I think you can rebase your branch and open a new PR based on it.
